Support for Cosign 2.0

vishal-chdhry opened the pull request:
Signed-off-by: Vishal Choudhary <sendtovishalchoudhary@gmail.com>
## Explanation

<!--
In a couple sentences, explain why this PR is needed and what it addresses. This should be an explanation a non-developer user can understand and covers the "why" question. It should also clearly indicate whether this PR represents an addition, a change, or a fix of existing behavior. This explanation will be used to assist in the release note drafting process.

THIS IS MANDATORY.
-->

## Related issue

Closes #6401

Upgrades to Cosign 2.0

<!--
Please link the GitHub issue this pull request resolves in the format of `Closes #1234`. If you discussed this change
with a maintainer, please mention her/him using the `@` syntax (e.g. `@JimBugwadia`).

If this change neither resolves an existing issue nor has sign-off from one of the maintainers, there is a
chance substantial changes will be requested or that the changes will be rejected.

You can discuss changes with maintainers in the [Kyverno Slack Channel](https://kubernetes.slack.com/).
-->

## Milestone of this PR
<!--

Add the milestone label by commenting `/milestone 1.2.3`.

-->

## What type of PR is this

<!--

> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading white spaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
-->

## Proposed Changes

<!--
Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. 

***NOTE***: If this PR results in new or altered behavior which is user facing, you **MUST** read and follow the steps outlined in the [PR documentation guide](pr_documentation.md) and add Proof Manifests as defined below.
-->

### Proof Manifests

<!--
Read and follow the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) for more details first. This section is for pasting your YAML manifests (Kubernetes resources and Kyverno policies) and Kyverno CLI test manifests which allow maintainers to prove the intended functionality is achieved by your PR. Please use proper fenced code block formatting, for example:

# Kubernetes resource

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: roles-dictionary
  namespace: default
data:
  allowed-roles: "[\"cluster-admin\", \"cluster-operator\", \"tenant-admin\"]"
```

# Kyverno CLI test manifest (please see docs for latest manifest format at https://kyverno.io/docs/kyverno-cli/). See kyverno/policies for complete examples of all related test files.

```yaml
name: prepend-image-registry
policies:
  - prepend_image_registry.yaml
resources:
  - resource.yaml
variables: values.yaml
results:
  - policy: prepend-registry
    rule: prepend-registry-containers
    resource: mypod
    # if mutate rule
    patchedResource: patchedResource01.yaml
    kind: Pod
    result: pass
```
-->

## Checklist

<!--
Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of
them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.
-->

- [x] I have read the [contributing guidelines](https://github.com/kyverno/kyverno/blob/main/CONTRIBUTING.md).
- [ ] I have read the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) and followed the process including adding proof manifests to this PR.
- [ ] This is a bug fix and I have added unit tests that prove my fix is effective.
- [ ] This is a feature and I have added CLI tests that are applicable.
- [ ] My PR needs to be cherry picked to a specific release branch which is <replace>.
- [ ] My PR contains new or altered behavior to Kyverno and
  - [ ] CLI support should be added and my PR doesn't contain that functionality.
  - [ ] I have added or changed [the documentation](https://github.com/kyverno/website) myself in an existing PR and the link is:
  <!-- Uncomment to link to the PR -->
  <!-- https://github.com/kyverno/website/pull/123 -->
  - [ ] I have raised an issue in [kyverno/website](https://github.com/kyverno/website) to track the documentation update and the link is:
  <!-- Uncomment to link to the issue -->
  <!-- https://github.com/kyverno/website/issues/1 -->

## Further Comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution
you did and what alternatives you considered, etc...
-->

Vishal-Chdhry commented:
After updating, the following test is failing with error `expected a signed timestamp to verify an expired certificate`
https://github.com/kyverno/kyverno/blob/ccb6da143a8810e417ea4b642b3b054e96d49e4a/pkg/cosign/cosign_test.go#L95-L96

The error originates here: https://github.com/sigstore/cosign/blob/f21081a1820939cddc41c1c119632a6c489d57fd/pkg/cosign/verify.go#L744-L754C3

> By default, artifact signatures will be uploaded to Rekor, for both key-based and identity-based signing. To not upload to Rekor, include --tlog-upload=false.

Since, artifacts signatrues are uploaded to Rekor in cosign 2.0, if I don't set `IgnoreTlog` to `true`, this part will throw and error as Rekor is not provided: https://github.com/sigstore/cosign/blob/f21081a1820939cddc41c1c119632a6c489d57fd/pkg/cosign/verify.go#L982-L984

The error mentioned in the beginning only depends on `IgnoreTlops` as `acceptableRFC3161Time` will always be nil as we are not providing TSA Certificates: https://github.com/sigstore/cosign/blob/f21081a1820939cddc41c1c119632a6c489d57fd/pkg/cosign/verify.go#L623-L631

To fix this, I believe we will either have to provide a cert that is not expired or provide a rekor url

JimBugwadia commented:
@Vishal-Chdhry - the Rekor address is optional in Kyverno policy rules. 

Can we set `IgnoreTlog` to `true` when `rekor` is set to `nil`?

Vishal-Chdhry commented:
> Can we set IgnoreTlog to true when rekor is set to nil?

@JimBugwadia Yes, I have updated that in the `buildCosignOptions`

But in the new version they have added checks for expired certificates here: [sigstore/cosign@f21081a/pkg/cosign/verify.go#L744-L754C3](https://github.com/sigstore/cosign/blob/f21081a1820939cddc41c1c119632a6c489d57fd/pkg/cosign/verify.go#L744-L754C3)

This part throws an error in the tests, We will have to update the test image with a fresh signature

JimBugwadia commented:
> But in the new version they have added checks for expired certificates here: [sigstore/cosign@f21081a/pkg/cosign/verify.go#L744-L754C3](https://github.com/sigstore/cosign/blob/f21081a1820939cddc41c1c119632a6c489d57fd/pkg/cosign/verify.go#L744-L754C3)
> This part throws an error in the tests, We will have to update the test image with a fresh signature

Ok.  Its hard to tell which test is failing due to that change, as there are other errors in your test run.  Can you please fix this?

```
Error: ../../../go/pkg/mod/github.com/in-toto/in-toto-golang@v0.8.0/in_toto/model.go:1079:21: s.signer.Verify undefined (type *dsse.EnvelopeSigner has no field or method Verify)
```

Vishal-Chdhry commented:
@JimBugwadia there was some issue with the in-toto pkg which got fixed when i updated the version to 0.9.

Here is the error that i get now
```
--- FAIL: TestCosignKeyless (7.66s)
    kyverno/pkg/cosign/cosign_test.go:92: assertion failed: expected error to contain "subject mismatch: expected jim, received jim@nirmata.com", got "no matching signatures:\nexpected a signed timestamp to verify an expired certificate\n expected a signed timestamp to verify an expired certificate"
FAIL
FAIL	github.com/kyverno/kyverno/pkg/cosign	8.529s
FAIL
```
which says `expected a signed timestamp to verify an expired certificate`

JimBugwadia commented:
So, for the `TestCosignKeyless` test are we using the public Rekor?  You may need to set `opts. RekorURL` so it fetches the certs from the public tlog.

JimBugwadia commented:
@Vishal-Chdhry - please address the DCO errors.

JimBugwadia commented on pkg/cosign/cosign.go:
```diff
@@ -181,11 +180,23 @@ func buildCosignOptions(ctx context.Context, opts images.Options) (*cosign.Check
 		}
 	}
 
+	if opts.RekorURL == "" && opts.Key == "" {
```
Should this be only for Keyless or all attestor types?

Vishal-Chdhry replied:
When I allow it for Keyed, it gives this error
```
I0525 05:09:25.555208       1 cosign.go:61] cosign "msg"="image verification failed" "error"="no matching signatures:\nsignature not found in transparency log\n searching log query: [POST /api/v1/log/entries/retrieve][400] searchLogQueryBadRequest  &{Code:400 Message:verifying signature: invalid signature when validating ASN.1 encoded signature}\n searching log query: [POST /api/v1/log/entries/retrieve][400] searchLogQueryBadRequest  &{Code:400 Message:verifying signature: invalid signature when validating ASN.1 encoded signature}\n searching log query: [POST /api/v1/log/entries/retrieve][400] searchLogQueryBadRequest  &{Code:400 Message:verifying signature: invalid signature when validating ASN.1 encoded signature}"
```

`no matching signatures:\nsignature not found in transparency log\n searching log query: [POST /api/v1/log/entries/retrieve` - This part looks related to rekor, thats why I added it just for keyless.

Log for [keyed-basic](https://github.com/kyverno/kyverno/tree/main/test/conformance/kuttl/verifyImages/clusterpolicy/standard/keyed-basic) kuttl test - https://pastebin.com/KxMnqB5s

JimBugwadia commented on pkg/cosign/cosign.go:
```diff
@@ -181,11 +180,23 @@ func buildCosignOptions(ctx context.Context, opts images.Options) (*cosign.Check
 	}
 
+	if opts.RekorURL == "" && opts.Key == "" {
+		// KEYLESS Verification and RekorURL is not provided
```
Use sentence case for comments:

KEYLESS Verification -> Keyless verification

JimBugwadia commented on pkg/cosign/cosign.go:
```diff
@@ -181,11 +180,23 @@ func buildCosignOptions(ctx context.Context, opts images.Options) (*cosign.Check
+
+		cosignOpts.RekorPubKeys, err = cosign.GetRekorPubs(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load RekorPubKeys: %w", err)
```
"failed to load RekorPubKeys" -> "failed to load Rekor public keys"

JimBugwadia commented on pkg/cosign/mock.go:
```diff
@@ -53,17 +54,31 @@ func (m *mock) getSignatures(signedImgRef name.Reference) ([]oci.Signature, bool
-
-func (s *sig) Payload() ([]byte, error) {
-	return s.cosignPayload.Payload, nil
+func GetOCISignatureFromCosignSignedPayload(sp cosign.SignedPayload) (oci.Signature, error) {
```
Shorten `GetOCISignatureFromCosignSignedPayload` to `GetSignatureFromSignedPayload` or just `GetSignature`.

No need to repeat package names. Also the params indicate we are passing a signed payload and getting back a signature, so the "From..." is optional.

JimBugwadia replied:
Is this method used outside the package? If not please make it private / lowecase.

JimBugwadia commented on pkg/cosign/registryOptions.go:
```diff
@@ -0,0 +1,79 @@
+package cosign
```
Why not use https://github.com/sigstore/cosign/blob/65eb28af970d133adeefdc6c48d6e9304dd8cc3a/cmd/cosign/cli/options/registry.go#L88?

Vishal-Chdhry replied:
I decided against using it because it is in the CLI package, and since we are trying to remove CLI packages I tried to create our own. There is no package that provides it other than the CLI package. Also the function used here is quite similar to the notary authentication, I am thinking about making a common package that will fetch gcr remote options for both of them

JimBugwadia replied:
Ok. Should we change this now, or wait till we remove other Cosign CLI dependencies?

Vishal-Chdhry replied:
This PR removes all the other CLI dependencies as well

Vishal-Chdhry commented:
I have updated the cosign verification flow to the way it is done in cosign cli.
```bash
cosign verify --key cosign.pub ghcr.io/kyverno/test-verify-image:signed --insecure-ignore-tlog=true  
```

```bash
cosign verify ghcr.io/jimbugwadia/pause2 --certificate-oidc-issuer  https://github.com/login/oauth --certificate-identity jim@nirmata.com --insecure-ignore-sct=true
```

I have added the option to set `ignore-tlog` and `ignore-sct`, and added a default rekor url in all cases just like in cosign 2.0 CLI.

So that the flow exactly matches the flow in the CLI and users can directly convert their `cosign verify` command to Kyverno image verification policy

I have also added rekor pubkeys so that user can set set custom rekor pubkeys, when not provided it will create a default one using `cosign.GetRekorPubs`

JimBugwadia commented:
@Vishal-Chdhry - please see this prior issue for SCT verification: https://github.com/kyverno/kyverno/issues/5165

JimBugwadia commented on api/kyverno/v1/image_verification_types.go:
```diff
@@ -226,9 +226,23 @@ type KeylessAttestor struct {
 type CTLog struct {
 	// URL is the address of the transparency log. Defaults to the public log https://rekor.sigstore.dev.
-	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Optional
```
In the various attestators, we allow `Rekor` to be `nil. 

What is the behavior when `Rekor` is set to `nil` in an attestor?

The prior behavior is documented in the comments, which may need to be updated.

```go
	// Rekor provides configuration for the Rekor transparency log service. If the value is nil,
	// Rekor is not checked. If an empty object is provided the public instance of
	// Rekor (https://rekor.sigstore.dev) is used.

```

Vishal-Chdhry replied:
When the `Rekor` is set to nil, I am providing https://rekor.sigstore.dev/  as the rekor url https://github.com/kyverno/kyverno/pull/7248/files#diff-4e334a7b5c4dc63f280d7a0e49015d54a381c6445fdc4db8065cc9c9d330e255R504-R506

JimBugwadia replied:
This is the opposite of the prior behavior....and may be confusing. 

Previously setting to `nil` meant that Rekor is not used. Now setting to `nil` means that the public Rekor is used, but if we inspect the policy the URL is not visible.

Perhaps we can make `Rekor` a required attribute and set the URL to the public instance by default, so it will be clear?

This will be a breaking change, so we will need to release note for users with existing policies.

JimBugwadia commented on api/kyverno/v1/image_verification_types.go:
```diff
@@ -226,9 +226,23 @@ type KeylessAttestor struct {
+
+	// RekorPubKey is an optional PEM encoded public key to use for a custom Rekor.
+	// +kubebuilder:validation:Optional
+	RekorPubKey string `json:"pubkey,omitempty" yaml:"pubkey,omitempty"`
```
Previously, one use case we supported was to verify a keyless attestation or image signature using the root certificate of Rekor.

Will this still be supported? May be good to add a `kutt` test case for this.

Vishal-Chdhry replied:
In the current behavior, we are providing the public rekor url and its public key by default, to test `RekorPubKey` with a kuttl test, I believe we will have to setup an instance of rekor for test.

So, I created this unit test (https://github.com/kyverno/kyverno/pull/7248/commits/497646a5c03766ef5805fcab54862c0451f7071a) where we:
1. Pass an invalid rekor url so it doesn't use the default rekor URL.
2. Pass an invalid rekor public key so that it throws an error.
3. Pass the right rekor public key and the test passes.

It tests that we can verify signatures only using the rekor public key

JimBugwadia replied:
That is not the same. The requirement was to allow the user to not specify Rekor and only provide a root CA.

JimBugwadia replied:
Can they specify `ignoreTlog` to achieve this?

vishal-chdhry replied:
> That is not the same. The requirement was to allow the user to not specify Rekor and only provide a root CA.

@JimBugwadia I am not sure if I completely understand this.

- If a user want to verify just the signature and not rekor entry, then they can set ignoreTlogs  for that
- If the user has the rekor root CA, they can just provide that CA and not the rekor url, and cosign will do the verification based on that rekor URL

Assuming that the root CA mentioned by you is the Rekor Root CA, for a keyless verification, the user can provide
1. Identity and Issuer
2. Rekor Root CA

This will be enough to verify the signature and the tlog

realshuting commented on go.mod:
```diff
@@ -2,6 +2,8 @@ module github.com/kyverno/kyverno
 
 go 1.20
 
+replace github.com/google/go-containerregistry/0_14 => github.com/google/go-containerregistry v0.14.1-0.20230425172351-b7c6e9dc3944
```
👍

You may want to combine this with other replaced packages, I'm seeing a conflict in this file.

realshuting commented:
Another linter issue and conflicts to fix:)

realshuting commented:
We need to resolve conflicts and get this merged for 1.11. Is there any pending item @vishal?

@eddycharly @JimBugwadia - could you review the PR once again?

eddycharly commented on api/kyverno/v1/image_verification_types.go:
```diff
@@ -243,6 +243,21 @@ type CTLog struct {
+
+	// IgnoreSCT requires that a certificate contain an embedded SCT during verification. An SCT is proof of inclusion in a certificate transparency log.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=false
```
this marker is not needed

eddycharly commented on api/kyverno/v1/image_verification_types.go:
```diff
@@ -243,6 +243,21 @@ type CTLog struct {
+
+	// IgnoreTlog skip tlog verification
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=false
```
this marker is not needed

eddycharly commented on pkg/notary/notary_test.go:
```diff
@@ -4,18 +4,18 @@ import (
 	"context"
 	"testing"
 
+	gcrremote "github.com/google/go-containerregistry/0_14/pkg/v1/remote" // TODO: Remove this once we upgrade tp cosign version < 2.0.2
```
what's this ?

Vishal-Chdhry replied:
While upgrading to `Cosign 2.1.1`,  I faced this issue with k8s-manifest-sigstore
```
github.com/sigstore/k8s-manifest-sigstore@v0.5.0/pkg/k8smanifest/provenance.go:569:32: undefined: rekorutil.PublicKey
```
I believe that is because `Cosign 2.0.2` uses `Rekor 1.1` and `Cosign 2.1.1 `uses `Rekor 1.2`

Current notary implementation uses gcr remote's Reuse function and cosign 2.0.2 doesn't allow us to upgrade to the latest version of gcr remote. So I am maintaining 2 versions of gcr remote, unit we can to `Cosign 2.1.1` to support it.

I have asked the maintainers of k8s-manifest-sigstore about it https://sigstore.slack.com/archives/C01DGF0G8U9/p1688724765487689

eddycharly replied:
Ok, we need to track this

eddycharly commented on pkg/cosign/registryOptions.go:
```diff
@@ -0,0 +1,79 @@
+	return opts, nil
+}
+
+func (o *RegistryOptions) GetRegistryClientOpts(ctx context.Context) []remote.Option {
```
aren't we using the options from the registry client ?

Vishal-Chdhry replied:
I will update it to the use `GCRRemoteOpts` once Amit's PR is merged: https://github.com/kyverno/kyverno/pull/7790

eddycharly replied:
Is there no other option ?

Vishal-Chdhry replied:
Should I remove this function after https://github.com/kyverno/kyverno/pull/7806 is merged? As this keychain will be overwritten anyways by `BuildRemoteOption`

eddycharly replied:
The idea is that we use the same keychain as the registry client.

Vishal-Chdhry replied:
@eddycharly I removed the registryOptions as client.go will overwrite the registry options anyways, and I added the missing parts to `client.go` in this commit: [`7d675bc` (#7248)](https://github.com/kyverno/kyverno/pull/7248/commits/7d675bce35c1a5c97a76a1bbe0d7b22923c486f1)

eddycharly commented on pkg/cosign/cosign.go:
```diff
@@ -178,11 +178,21 @@ func buildCosignOptions(ctx context.Context, opts images.Options) (*cosign.Check
-			return nil, fmt.Errorf("failed to create Rekor client from URL %s: %w", opts.RekorURL, err)
-		}
+	cosignOpts.IgnoreTlog = opts.IgnoreTlog
+	cosignOpts.RekorClient, err = rekorclient.GetRekorClient(opts.RekorURL)
```
would be nice to support tracing here

eddycharly commented on pkg/cosign/registryOptions.go:
```diff
@@ -0,0 +1,79 @@
+
+type RegistryOptions struct {
+	AllowInsecure      bool
+	AllowHTTPRegistry  bool
```
is this used somewhere ?

eddycharly commented on pkg/cosign/registryOptions.go:
```diff
@@ -0,0 +1,79 @@
+	AllowInsecure      bool
+	AllowHTTPRegistry  bool
+	KubernetesKeychain bool
+	TagPrefix          string
```
is this used somewhere ?

chipzoller commented:
Conflicts to resolve.

JimBugwadia commented on api/kyverno/v1/image_verification_types.go:
```diff
@@ -183,7 +183,7 @@ type StaticKeyAttestor struct {
 
 	// Rekor provides configuration for the Rekor transparency log service. If the value is nil,
-	// Rekor is not checked. If an empty object is provided the public instance of
+	// or an empty object is provided, the public instance of
```
@vishal-chdhry - so can a user specify that Rekor should not be used?

JimBugwadia replied:
Can they specify `ignoreTlog` to achieve this?

vishal-chdhry replied:
Yes, If a user set `ignoreTlog` to be true, rekor will not be used

realshuting commented:
Great work folks🎉
